server/src/main/java/org/elasticsearch/cluster/metadata/IndexReshardingMetadata.java

```diff
@@ -0,0 +1,315 @@
+ *   - If a source shard may still contain data for any target shard then it is in state `SOURCE`.
+ *   - When all targets for a source have moved to `SPLIT` (see below), then the source deletes all documents from
+ *     its store that are now the responsibility of the target shards and transitions to `DONE`.
+ *   This isn't strictly required to be persistent for correctness, but it can save time on recovery
+ *   by allowing a DONE shard to skip interrogating targets and repeating cleanup.
+ * * For each target shard, its current target state, one of `CLONE`, `HANDOFF`, `SPLIT`, or `DONE`.
+ *   - If the target has not yet copied all data from the source shard, then it is in `CLONE`.
+ *   - It moves to `HANDOFF` when it has copied all of its data from the source to indicate that it is now ready to
+ *     receive indexing actions, and starts RUNNING. After this point, the source may no longer contain the entire contents
+ *     of the target and must not index documents belonging to the target. But since search shards can't start up until
+ *     their corresponding index nodes are RUNNING, search requests would fail if they routed to the target shard immediately
+ *     after handoff. So at HANDOFF, the source shards continue to service searches, but block refresh since they cannot
+ *     be guaranteed to have seen documents indexed after HANDOFF.
+ *   - When the target shard's corresponding search replica has started running, the target requests that the source filter
+ *     search results belonging to the target, and moves the target shard's state moves to `SPLIT`. The target's search replica
+ *     likewise filters documents not belonging to the target, which may be present due to the target bootstrapping by copying
+ *     the source's lucene files.
+ *   - Upon entering `SPLIT`, the target starts deleting all documents from its lucene store that do not belong to it. When that
+ *     is complete, it moves to `DONE` and removes filters for other shards, which are no longer necessary.
+ *
+ * Note that each target shard's split operates independently and all may happen concurrently.
+ *
+ * When all source shards have transitioned to `DONE`, the resize is complete and this metadata may be removed from cluster state.
+ * We only allow at most a single resharding operation to be in flight for an index, so removing this metadata is a prerequisite
+ * to beginning another resharding operation.
+ */
+public record IndexReshardingMetadata(
+    int oldShardCount,
+    int newShardCount,
+    SourceShardState[] sourceShardStates,
+    TargetShardState[] targetShardStates
+) implements ToXContentFragment, Writeable {
+    public enum SourceShardState implements Writeable {
+        SOURCE,
+        DONE;
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            out.writeEnum(this);
+        }
+    }
+
+    public enum TargetShardState implements Writeable {
```

Comment:
I wonder if we need the "State" to be more like a Reshard enum (SOURCE (SOURCE - maybe we need to change this name for state / DONE), TARGET (CLONE, HANDOFF, SPLIT, DONE)) itself with variable characteristics on it. Similar to `RecoverySource` . 

And then it is:
- old [SOURCE (state: PRE_DELETE), SOURCE (state: PRE_DELETE)]
- new [TARGET (state: CLONE), TARGET (state: HANDOFF)]

Which then allows for an additional types once we implement shrink that can be serialized into the same old / new arrays?

Just something to think about before we commit to an initial serialization.

Reply:
I will think about it for a bit. It'd be nice to avoid a situation where an array could hold both SPLIT(SOURCE) and SHRINK(SOURCE) elements at the same time. I wonder whether we might want to wrap the whole array pair in an operation or something.

Reply:
I think we might get the extensibility you want and the type safety I want by adding an "operation" enum (SPLIT, SHRINK) before the shard arrays in the serialized form. On parsing the operation we could branch to parsing the remainder of the serialized metadata according to the operation. I'll give it a try, assuming that works for you?

Reply:
I don't have a strong opinion about how it is designed. I just suspect:

```
SplitSource
SplitTarget

ShrinkSource
ShrinkTarget
```

Will have different states associated with them so we might think about how we represent that. An enum for the operation type would also provide that (I assume).

Reply:
alright, makes sense, I'll play with a top level field to control how the rest of the resharding state is interpreted.

Reply:
I've made the IndexReshardingMetadata field into something like a tagged union in 7f6c7b9. To be honest, it felt like I was fighting the XContent model a little bit doing it, so it could be I'm missing something, but it does appear to work. I tried to find other examples of XContent that held versioned data but didn't see much applicable. Thoughts appreciated here.

Reply:
I've simplified this in 272e8f3